src/video/SDL_clipboard.c

```diff
@@ -18,10 +18,102 @@
      misrepresented as being the original software.
   3. This notice may not be removed or altered from any source distribution.
 */
+#include "SDL_error.h"
 #include "SDL_internal.h"
 
 #include "SDL_sysvideo.h"
 
+SDL_ClipboardData *
+SDL_CreateClipboardData(void *data, size_t len, const char *mime_type)
+{
+    SDL_ClipboardData *cb = SDL_calloc(1, sizeof(SDL_ClipboardData));
+    if (cb == NULL) {
+        SDL_OutOfMemory();
+        return NULL;
+    }
+
+    cb->data = SDL_calloc(1, len);
+    if (cb->data == NULL) {
+        SDL_OutOfMemory();
+        return NULL;
+    }
+    SDL_memcpy(cb->data, data, len);
+
+    cb->mime_type = SDL_strdup(mime_type);
+    if (cb->mime_type == NULL) {
+        SDL_OutOfMemory();
+        return NULL;
+    }
+    cb->size = len;
+
+    return cb;
+}
+
+int
+SDL_AddClipboardData(SDL_ClipboardData *cbdata, void *data, size_t len, const char *mime_type)
+{
+    SDL_ClipboardData *tail;
+
+    if (cbdata == NULL) {
+        return SDL_InvalidParamError("cbdata");
+    }
+
+    while (cbdata->next != NULL) {
+        cbdata = cbdata->next;
+    }
+
+    tail = SDL_CreateClipboardData(data, len, mime_type);
+    if (tail == NULL) {
+        /* Error is set in the create function */
+        return -1;
+    }
+
+    cbdata->next = tail;
+    return 0;
+}
```

Comment:
Regarding the two above functions. I see 3 options for SDL_ClipboardData.
 1. Keep as is and expose to users
 2. Hide the struct in the SDL_VideoDevice and don't expose to user
 3. Push functions all the way down to driver and let the driver chose how to handle Add and Set calls. (I like this the least because it will likely result in a lot of "similar" code in drivers.)